File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/events/EcoEventManager.java
package com.willfp.eco.internal.events;

import com.willfp.eco.core.events.EventManager;
import com.willfp.eco.core.PluginDependent;
import com.willfp.eco.core.EcoPlugin;
import org.bukkit.Bukkit;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

public class EcoEventManager extends PluginDependent<EcoPlugin> implements EventManager {
    @ApiStatus.Internal
    public EcoEventManager(@NotNull final EcoPlugin plugin) {
        super(plugin);
    }

    @Override
    public void registerListener(@NotNull final Listener listener) {
        Bukkit.getPluginManager().registerEvents(listener, this.getPlugin());
    }

    @Override
    public void unregisterListener(@NotNull final Listener listener) {
        HandlerList.unregisterAll(listener);
    }

    @Override
    public void unregisterAllListeners() {
        HandlerList.unregisterAll(this.getPlugin());
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/events/EcoEventManager.kt
package com.willfp.eco.internal.events

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.events.EventManager
import com.willfp.eco.core.packet.PacketEvent
import com.willfp.eco.core.packet.PacketListener
import com.willfp.eco.core.packet.PacketPriority
import org.bukkit.Bukkit
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener


private class RegisteredPacketListener(
    val plugin: EcoPlugin,
    val listener: PacketListener
)

private val listeners = mutableMapOf<PacketPriority, MutableList<RegisteredPacketListener>>()

fun PacketEvent.handleSend() {
    for (priority in PacketPriority.values()) {
        for (listener in listeners[priority] ?: continue) {
            try {
                listener.listener.onSend(this)
            } catch (e: Exception) {
                listener.plugin.logger.warning(
                    "Exception in packet listener ${listener.listener.javaClass.simpleName}" +
                            " for packet ${packet.javaClass.simpleName}!"
                )
                e.printStackTrace()
            }
        }
    }
}

fun PacketEvent.handleReceive() {
    for (priority in PacketPriority.values()) {
        for (listener in listeners[priority] ?: continue) {
            try {
                listener.listener.onReceive(this)
            } catch (e: Exception) {
                listener.plugin.logger.warning(
                    "Exception in packet listener ${listener.listener.javaClass.simpleName}" +
                            " for packet ${packet.javaClass.simpleName}!"
                )
                e.printStackTrace()
            }
        }
    }
}

class EcoEventManager(private val plugin: EcoPlugin) : EventManager {
    override fun registerListener(listener: Listener) {
        Bukkit.getPluginManager().registerEvents(listener, plugin)
    }

    override fun unregisterListener(listener: Listener) {
        HandlerList.unregisterAll(listener)
    }

    override fun unregisterAllListeners() {
        HandlerList.unregisterAll(plugin)
        for (value in listeners.values) {
            value.removeIf { it.plugin == plugin }
        }
    }

    override fun registerPacketListener(listener: PacketListener) {
        listeners.getOrPut(listener.priority) { mutableListOf() }.add(
            RegisteredPacketListener(
                plugin,
                listener
            )
        )
    }
}
